wallet-api/src/Wallet/API.hs

```diff
@@ -100,76 +104,97 @@ $(deriveEq1 ''EventTriggerF)
+
+instance Corecursive (EventTrigger a) where
+    embed (Compose (v, t)) = let f (a, b) = EventTrigger . Fix . AnnTriggerF . Compose $ (a, b) in
+        case t of
+            And (EventTrigger l) (EventTrigger r) -> f (v, And l r)
+            Or (EventTrigger l) (EventTrigger r)  -> f (v, Or l r)
+            Not (EventTrigger b)                  -> f (v, Not b)
+            PAlways                               -> f (v, PAlways)
+            PNever                                -> f (v, PNever)
+            BlockHeightRange r                    -> f (v, BlockHeightRange r)
+            FundsAtAddress addr r                 -> f (v, FundsAtAddress addr r)
 
 -- | `andT l r` is true when `l` and `r` are true.
-andT :: EventTrigger -> EventTrigger -> EventTrigger
-andT l = embed . And l
+andT :: EventTrigger () -> EventTrigger () -> EventTrigger ()
+andT l = embed . Compose . pure . And l
 
 -- | `orT l r` is true when `l` or `r` are true.
-orT :: EventTrigger -> EventTrigger -> EventTrigger
-orT l = embed . Or l
+orT :: EventTrigger () -> EventTrigger () -> EventTrigger ()
+orT l = embed . Compose . pure . Or l
 
 -- | `alwaysT` is always true
-alwaysT :: EventTrigger
-alwaysT = embed PAlways
+alwaysT :: EventTrigger ()
+alwaysT = embed . Compose . pure $ PAlways
 
 -- | `neverT` is never true
-neverT :: EventTrigger
-neverT = embed PNever
+neverT :: EventTrigger ()
+neverT = embed . Compose . pure $ PNever
 
 -- | `blockHeightT r` is true when the block height is in the range `r`
-blockHeightT :: Range Height -> EventTrigger
-blockHeightT = embed . BlockHeightRange
+blockHeightT :: Range Height -> EventTrigger ()
+blockHeightT = embed . Compose . pure . BlockHeightRange
 
 -- | `fundsAtAddressT a r` is true when the funds at `a` are in the range `r`
-fundsAtAddressT :: Address' -> Range Value -> EventTrigger
-fundsAtAddressT a = embed . FundsAtAddress a
+fundsAtAddressT :: Address' -> Range Value -> EventTrigger ()
+fundsAtAddressT a = embed . Compose . pure . FundsAtAddress a
 
 -- | `notT t` is true when `t` is false
-notT :: EventTrigger -> EventTrigger
-notT = embed . Not
+notT :: EventTrigger () -> EventTrigger ()
+notT = embed . Compose . pure . Not
 
 -- | Check if the given block height and UTXOs match the
 --   conditions of an [[EventTrigger]]
-checkTrigger :: Height -> Map.Map Address' Value -> EventTrigger -> Bool
-checkTrigger h mp = cata chk where
-    chk = \case
-        And l r -> l && r
-        Or  l r -> l || r
-        Not r   -> not r
-        PAlways -> True
-        PNever -> False
-        BlockHeightRange r -> h `inRange` r
+checkTrigger :: Height -> Map.Map Address' Value -> EventTrigger () -> Bool
+checkTrigger h mp = getAnnot . annTruthValue h mp
+
+-- | Annotate each node in an `EventTriggerF` with its truth value given a block height
+--   and a set of unspent outputs
+annTruthValue :: Height -> Map.Map Address' Value -> EventTrigger a -> EventTrigger Bool
```

Comment:
Then this would end up with type `... -> (Fix EventTriggerF) -> (Fix (Compose (AnnTriggerF a) EventTriggerF))` or something. With appropriate type synonyms.

Reply:
I see, the newtype isn't necessary at all. I'll remove it.